// @opengsn/relay/dist/StoredTransaction.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethUtils = __importStar(require("ethereumjs-util"));
var ServerAction;
(function (ServerAction) {
    ServerAction[ServerAction["REGISTER_SERVER"] = 0] = "REGISTER_SERVER";
    ServerAction[ServerAction["ADD_WORKER"] = 1] = "ADD_WORKER";
    ServerAction[ServerAction["RELAY_CALL"] = 2] = "RELAY_CALL";
    ServerAction[ServerAction["VALUE_TRANSFER"] = 3] = "VALUE_TRANSFER";
    ServerAction[ServerAction["DEPOSIT_WITHDRAWAL"] = 4] = "DEPOSIT_WITHDRAWAL";
    ServerAction[ServerAction["PENALIZATION"] = 5] = "PENALIZATION";
    ServerAction[ServerAction["SET_OWNER"] = 6] = "SET_OWNER";
})(ServerAction = exports.ServerAction || (exports.ServerAction = {}));
/**
 * Make sure not to pass {@link StoredTransaction} as {@param metadata}, as it will override fields from {@param tx}!
 * @param tx
 * @param metadata
 */
function createStoredTransaction(tx, metadata) {
    if (tx.to == null) {
        throw new Error('tx.to must be defined');
    }
    const details = {
        to: ethUtils.bufferToHex(tx.to.toBuffer()),
        gas: ethUtils.bufferToInt(tx.gasLimit.toBuffer()),
        gasPrice: ethUtils.bufferToInt(tx.gasPrice.toBuffer()),
        data: ethUtils.bufferToHex(tx.data),
        nonce: ethUtils.bufferToInt(tx.nonce.toBuffer()),
        txId: ethUtils.bufferToHex(tx.hash()),
        value: ethUtils.bufferToHex(tx.value.toBuffer())
    };
    return Object.assign({}, details, metadata);
}
exports.createStoredTransaction = createStoredTransaction;
